File: core/src/main/java/com/terralite/core/registry/ResourceId.java
package com.terralite.core.registry;

import java.util.Objects;
import java.util.regex.Pattern;

public record ResourceId(String namespace, String path) {
    private static final Pattern VALID_NAMESPACE = Pattern.compile("[a-z0-9_.-]+");
    private static final Pattern VALID_PATH = Pattern.compile("[a-z0-9_./-]+");

    public ResourceId {
        Objects.requireNonNull(namespace, "namespace");
        Objects.requireNonNull(path, "path");

        if (!VALID_NAMESPACE.matcher(namespace).matches()) {
            throw new IllegalArgumentException("Invalid namespace: " + namespace);
        }

        if (!VALID_PATH.matcher(path).matches()) {
            throw new IllegalArgumentException("Invalid path: " + path);
        }
    }

    public static ResourceId of(String namespace, String path) {
        return new ResourceId(namespace, path);
    }

    public static ResourceId id(String value) {
        return parse(value);
    }

    public static ResourceId parse(String value) {
        Objects.requireNonNull(value, "value");

        int separator = value.indexOf(':');
        if (separator <= 0 || separator == value.length() - 1 || value.indexOf(':', separator + 1) != -1) {
            throw new IllegalArgumentException("Invalid resource id: " + value);
        }

        return new ResourceId(value.substring(0, separator), value.substring(separator + 1));
    }

    @Override
    public String toString() {
        return namespace + ":" + path;
    }
}

File: core/src/commonMain/kotlin/com/terralite/core/registry/ResourceId.kt
package com.terralite.core.registry

/**
 * A namespaced identifier, e.g. `terralite:blocks/stone`.
 *
 * Both components must be lowercase alphanumeric with `_`, `.`, `/`, or `-`
 * (namespace also allows `-`; path also allows `/`).
 */
@JvmRecord
data class ResourceId(val namespace: String, val path: String) {

    init {
        require(VALID_NAMESPACE.matches(namespace)) { "Invalid namespace: $namespace" }
        require(VALID_PATH.matches(path)) { "Invalid path: $path" }
    }

    override fun toString(): String = "$namespace:$path"

    companion object {
        private val VALID_NAMESPACE = Regex("[a-z0-9_.-]+")
        private val VALID_PATH = Regex("[a-z0-9_./-]+")

        @JvmStatic
        fun of(namespace: String, path: String): ResourceId = ResourceId(namespace, path)

        /** Alias for [parse] — matches the Java API shorthand `ResourceId.id("ns:path")`. */
        @JvmStatic
        fun id(value: String): ResourceId = parse(value)

        @JvmStatic
        fun parse(value: String): ResourceId {
            val sep = value.indexOf(':')
            require(sep > 0 && sep < value.length - 1 && value.indexOf(':', sep + 1) == -1) {
                "Invalid resource id: $value"
            }
            return ResourceId(value.substring(0, sep), value.substring(sep + 1))
        }
    }
}
